### PR TITLE
bgpv1: Set nexthop of locally originated routes to 0.0.0.0 (or ::)

### PR DESCRIPTION
**Pull request by YutaroHayakawa**
Please see the commit message and comments on the code for more details.

```release-note
bgpv1: Use IP address used for peering as a nexthop
```

### CODE REVIEW

**Comment by ldelossa** [comment]
@YutaroHayakawa this change makes sense to me. 

Brainstorm with me for a second, is there anything about iBGP vs eBGP next hop rules which we need to concern ourselves with here?

**Comment by YutaroHayakawa** [comment]
RFC section below is the ground truth (they also apply for MP-BGP MP-REACH-NLRI), but usually, what we need to care about iBGP vs. eBGP is

- iBGP doesn't modify next-hop
- eBGP always modifies next-hop to self IP address

There are some exceptions, but I believe that advertising with Node's local IP address should work for locally originated routes like PodCIDR or LBVIP. And I don't come up with any other use cases we redistribute the routes got from another routing protocol (e.g. OSPF, ISIS, etc...) for the moment.

https://datatracker.ietf.org/doc/html/rfc4271#section-5.1.3

**Comment by ldelossa** [comment]
@YutaroHayakawa thanks for pointing me to the relevant spec section. After a quick read I see the changes here are good. 

There's a few reasons the PR is not mergable. Take a look at the lint errors in reconcile_test.go

**Comment by YutaroHayakawa** [comment]
I should add an RFC reference to the comment. Let me do that.

>There's a few reasons the PR is not mergable. Take a look at the lint errors in reconcile_test.go

**Comment by YutaroHayakawa** [comment]
Travis failure is hitting this: https://app.travis-ci.com/github/cilium/cilium/jobs/567066920

Let me retry.

**Comment by YutaroHayakawa** [comment]
@ldelossa All fixes are in. Please review when you have time.

**Review by tklauser — APPROVED** [review]
LGTM, contingent on @ldelossa's approval.

File: pkg/bgpv1/gobgp/server.go [hunk 1]
```diff
@@ -165,32 +164,44 @@ func (sc *ServerWithConfig) RemoveNeighbor(ctx context.Context, n *v2alpha1api.C
...
-			return Advertisement{}, fmt.Errorf("cannot advertise IPv4 path %v, control plane has no configured IPv4 address for next hop", ip.String())
-		}
 		prefixLen, _ := ip.Mask.Size()
 		nlri, _ := apb.New(&gobgp.IPAddressPrefix{
 			PrefixLen: uint32(prefixLen),
 			Prefix:    ip.IP.String(),
 		})
+		// Currently, we only support advertising locally originated paths (the paths generated in Cilium
+		// node itself, not the paths received from another BGP Peer or redistributed from another routing
+		// protocols. In this case, the nexhop address should be the address used for peering. That means
```
Comment (ldelossa) [inline pkg/bgpv1/gobgp/server.go]: typo: s/nexhop/nexthop

File: pkg/bgpv1/gobgp/server.go [hunk 2]
```diff
@@ -165,32 +164,44 @@ func (sc *ServerWithConfig) RemoveNeighbor(ctx context.Context, n *v2alpha1api.C
...
+		// nexthop address varies depends on which neighbors we are  peering.
```
Comment (ldelossa) [inline pkg/bgpv1/gobgp/server.go]: Having a but of trouble understanding:

```
// protocols. In this case, the nexhop address should be the address used for peering. That means nexthop address varies depends on which neighbors we are  peering. 
```

also extra space between 'are' and 'peering'.

**Review by ldelossa — APPROVED** [review]
Just a few nits in the comments, but otherwise looks good.

File: pkg/bgpv1/gobgp/server.go (hunk 2 above)
Comment (YutaroHayakawa) [reply pkg/bgpv1/gobgp/server.go]: I fixed the comment. Now is it ok with you?

Comment (ldelossa) [reply pkg/bgpv1/gobgp/server.go]: yup

**Comment by YutaroHayakawa** [comment]
/test-1.16-4.9

**Comment by YutaroHayakawa** [comment]
/test-1.16-4.9

**Comment by YutaroHayakawa** [comment]
/test-1.21-5.4

**Comment by YutaroHayakawa** [comment]
/test-1.22-4.19

**Comment by YutaroHayakawa** [comment]
/test-1.23-net-next

**Comment by YutaroHayakawa** [comment]
Rebasing to the master to fix this issue https://github.com/cilium/cilium/pull/19523

**Comment by YutaroHayakawa** [comment]
Multicluster: https://github.com/cilium/cilium/issues/19126
ConformanceAKS: https://github.com/cilium/cilium/issues/18125

**Comment by YutaroHayakawa** [comment]
Reported these flakes.
ConformanceEKS: https://github.com/cilium/cilium/issues/19551
ConformanceGKE: https://github.com/cilium/cilium/issues/19552

**Comment by YutaroHayakawa** [comment]
Since this PR is not related to these tests, I think it's safe to make this ready to merge now.

**Comment by ldelossa** [comment]
Yup I think you're good @YutaroHayakawa

**Pull request merged**